### PR TITLE
Tighten generator validation

### PR DESCRIPTION
**Pull request by kargakis**
This is a way to validate generators, pushing it here to get some early feedback before going upstream.

@deads2k @fabianofranz @liggitt @smarterclayton 

@derekwaynecarr also since you have been using generators lately

### CODE REVIEW

**Comment by kargakis** [comment]
[test]

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory.go [hunk 1]
```diff
@@ -108,11 +110,15 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
 	flags := pflag.NewFlagSet("", pflag.ContinueOnError)
 	flags.SetNormalizeFunc(util.WarnWordSepNormalizeFunc) // Warn for "_" flags
 
-	generators := map[string]kubectl.Generator{
-		"run/v1":     kubectl.BasicReplicationController{},
-		"run-pod/v1": kubectl.BasicPod{},
-		"service/v1": kubectl.ServiceGeneratorV1{},
-		"service/v2": kubectl.ServiceGeneratorV2{},
+	generators := map[string]map[string]kubectl.Generator{
```
Comment (deads2k) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory.go]: If we're at the point of enumerating which generators are valid for which commands, does it start to make more sense to have a structured set of input arguments that are plumbed through?  This is essentially saying callers have to be aware of the shape of magic container they are sending in.

Comment (kargakis) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory.go]: I think a solution is to isolate the buckets and return them to the corresponding commands using separate calls. Generators vary a lot between them and having separate calls for separate groups makes sense to me.

Comment (kargakis) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory.go]: *updated*

**Comment by kargakis** [comment]
```
!!! Error in test/cmd/builds.sh:43
	'oc start-build --from-webhook="${webhook}"' exited with status 1
Call stack:
	1: test/cmd/builds.sh:43 main(...)
Exiting with status 1
!!! Error in hack/test-cmd.sh:281
	'${test}' exited with status 1
Call stack:
	1: hack/test-cmd.sh:281 main(...)
Exiting with status 1
[FAIL] !!!!! Test Failed !!!!
```

**Comment by kargakis** [comment]
^ @bparees @csrwng is this a known flake? ^

**Comment by csrwng** [comment]
No, but I did make a change to oc start-build --from-webhook yesterday. Looking into it.

**Comment by csrwng** [comment]
So prior to my fix, start-build --from-webhook was a no-op. I guess now it's actually trying to kick off a build and failing:

```
"Internal error occurred: could not generate a build: imageStreamTag \"ruby-20-centos7:latest\" not found",
```

**Comment by bparees** [comment]
@csrwng @kargakis https://github.com/openshift/origin/pull/5970

**Comment by csrwng** [comment]
@bparees :+1:

**Comment by bparees** [comment]
@csrwng apparently you were very (un)lucky your pull merged...  the only way it passes as it is right now is if the import manages to finish before the build is started.

**Comment by kargakis** [comment]
[test]

**Comment by kargakis** [comment]
https://github.com/kubernetes/kubernetes/pull/17516 is in upstream. 

Let's get this after the rebase to ease @soltysh's ~~torture~~ rebase:)

**Comment by soltysh** [comment]
> Let's get this after the rebase to ease @soltysh's torture rebase process:)

Thank you!

**Comment by danmcp** [comment]
@kargakis Bump

**Comment by kargakis** [comment]
Will get this in after the rebase lands

**Comment by kargakis** [comment]
This is partially superceded by `oc create route` work. We can remove the ugliness of flag validation once the next rebase is in.

**Pull request closed without merging**